src/bsp/src/stm32/src/InterlocManager.cpp

```diff
@@ -66,3 +82,174 @@ bool InterlocManager::constructUWBHeader(uint16_t destinationId,
+    double tof;
+    uint64_t tofDtu;
+
+    // receive poll message
+    device.receive(rxFrame, 0);
+
+    if (!isFrameGood(rxFrame, m_logger)) {
+        return 0;
+    }
+    if (rxFrame.m_frame->m_functionCode != UWBMessages::FunctionCode::TWR_POLL) {
+        return 0;
+    }
+
+    pollRxTs = rxFrame.m_rxTimestamp;
+
+    // construct and send Response message
+    constructUWBHeader(destinationId, UWBMessages::DATA, UWBMessages::TWR_RESPONSE,
+                       (uint8_t*)&responseMsg, sizeof(responseMsg));
+    //    respTxTs = (pollRxTs + (POLL_RX_TO_RESP_TX_DLY_UUS * UUS_TO_DWT_TIME))>>8;
+
+    device.transmit((uint8_t*)&responseMsg, sizeof(responseMsg));
+    //    device.transmitDelayed((uint8_t*)&responseMsg,sizeof(responseMsg),respTxTime);
+
+    // wait for Final message reception
+    device.receive(rxFrame, 0);
+    if (!isFrameGood(rxFrame, m_logger)) {
+        return 0;
+    }
+    if (rxFrame.m_frame->m_functionCode != UWBMessages::FunctionCode::TWR_FINAL) {
+        return 0;
+    }
+
+    UWBMessages::TWRFinal* finalTWRFrame = (UWBMessages::TWRFinal*)(rxFrame.m_frame);
+    device.getTxTimestamp(&respTxTs);
+    finalRxTs = rxFrame.m_rxTimestamp;
+
+    // evaluate distance
+    varA = (double)finalTWRFrame->m_respMinPoll;
+    varB = (double)((uint32_t)finalRxTs - (uint32_t)respTxTs);
+    varC = (double)(finalTWRFrame->m_finaleMinResp);
+    varD = (double)((uint32_t)respTxTs - (uint32_t)pollRxTs);
+
+    tofDtu = ((varA * varB - varC * varD) / (varA + varB + varC + varD));
+
+    tof = tofDtu * DWT_TIME_UNITS;
+    distanceEval = tof * SPEED_OF_LIGHT;
+
+    return distanceEval;
+}
+
+bool InterlocManager::sendTWRSequence(uint16_t destinationId, Decawave& device) {
+
+    UWBRxFrame responseFrame;
+    UWBRxFrame finalFrame;
+
+    UWBMessages::TWRPoll pollMsg{};
+    UWBMessages::TWRFinal finalMsg{};
+
+    uint64_t responseTimestamp;
+    uint64_t pollTimestamp;
+    uint64_t finalTxTime;
+    uint64_t finalTxTs;
+
+    // Construct poll message
+    constructUWBHeader(destinationId, UWBMessages::BEACON, UWBMessages::TWR_POLL,
+                       (uint8_t*)&pollMsg, sizeof(pollMsg));
+    device.transmitAndReceiveDelayed((uint8_t*)(&pollMsg), sizeof(UWBMessages::DWFrame),
+                                     POLL_TX_TO_RESP_RX_DLY_UUS, responseFrame,
+                                     RESP_RX_TIMEOUT_UUS);
+
+    if (!isFrameGood(responseFrame, m_logger)) {
+        return false;
+    }
+
+    if (responseFrame.m_frame->m_functionCode != UWBMessages::FunctionCode::TWR_RESPONSE) {
+        return false;
+    }
+
+    //  Retrieve poll transmission and response reception timestamp.
+    device.getTxTimestamp(&pollTimestamp);
+    responseTimestamp = responseFrame.m_rxTimestamp;
+
+    //  Compute final message transmission time
+    finalTxTime = (responseTimestamp + (RESP_RX_TO_FINAL_TX_DLY_UUS * UUS_TO_DWT_TIME)) >> 8;
+    finalTxTs = (finalTxTime << 8); //+device.getTxAntennaDLY();
```

Comment:
Remove the commented section if it is not to be used

Reply:
I prefer to let it there as a reminder that the delays are not properly set. More reading needs to be done in the following document : https://www.google.com/search?channel=fs&client=ubuntu&q=DecaRanging+ARM+Source+Code+Guide